engine/math/vector.lisp

```diff
@@ -0,0 +1,203 @@
+  (make-float4 (- (x vector1) (x vector2))  (- (y vector1) (y vector2))  (- (z vector1) (z vector2))  (- (w vector1) (w vector2))))
+
+
+(defmethod mul ((vector float2) scalar)
+  (make-float2 (* (x vector) scalar)  (* (y vector) scalar)))
+
+(defmethod mul ((vector float3) scalar)
+  (make-float3 (* (x vector) scalar)  (* (y vector) scalar)  (* (z vector) scalar)))
+
+(defmethod mul ((vector float4) scalar)
+  (make-float4 (* (x vector) scalar)  (* (y vector) scalar)  (* (z vector) scalar)  (* (w vector) scalar)))
+
+
+(defmethod div ((vector float2) scalar)
+  (make-float2 (/ (x vector) scalar)  (/ (y vector) scalar)))
+
+(defmethod div ((vector float3) scalar)
+  (make-float3 (/ (x vector) scalar)  (/ (y vector) scalar)  (/ (z vector) scalar)))
+
+(defmethod div ((vector float4) scalar)
+  (make-float4 (/ (x vector) scalar)  (/ (y vector) scalar)  (/ (z vector) scalar)  (/ (w vector) scalar)))
+
+
+(defmethod dot ((vector1 float2) vector2)
+  (+ (* (x vector1) (x vector2))  (* (y vector1) (y vector2))))
+
+(defmethod dot ((vector1 float3) vector2)
+  (+ (* (x vector1) (x vector2))  (* (y vector1) (y vector2)) (* (z vector1) (z vector2))))
+
+(defmethod dot ((vector1 float4) vector2)
+  (+ (* (x vector1) (x vector2))  (* (y vector1) (y vector2)) (* (z vector1) (z vector2)) (* (w vector1) (w vector2))))
+
+
+(defmethod length ((vector float2))
+  (sqrt (+ (expt (x vector) 2)  (expt (y vector) 2))))
+  
+(defmethod length ((vector float3))
+  (sqrt (+ (expt (x vector) 2)  (expt (y vector) 2) (expt (z vector) 2))))
+  
+(defmethod length ((vector float4))
+  (sqrt (+ (expt (x vector) 2)  (expt (y vector) 2) (expt (z vector) 2) (expt (w vector) 2))))
+
+
+(defmethod eq ((vector1 float2) vector2)
+  (and (= (x vector1) (x vector2)) (= (y vector1) (y vector2))))
+
+(defmethod eq ((vector1 float3) vector2)
+  (and (= (x vector1) (x vector2)) (= (y vector1) (y vector2)) (= (z vector1) (z vector2))))
+
+(defmethod eq ((vector1 float2) vector2)
+  (and (= (x vector1) (x vector2)) (= (y vector1) (y vector2)) (= (z vector1) (z vector2)) (= (w vector1) (w vector2))))
+
+
+(defun cross (vector1 vector2)
+  (make-float3 (- (* (y vector1) (z vector2)) (* (z vector1) (y vector2)))  (- (* (z vector1) (x vector2)) (* (x vector1) (z vector2)))  (- (* (x vector1) (y vector2)) (* (y vector1) (x vector2))))) 
+
+
+(defmethod normalize (vector)
+  (div vector (length vector)))
+
+
+(defmethod angle (vector1 vector2)
+  (acos (/ (dot vector1 vector 2) (* (length vector1) (length vector2)))))
+
+
+(defmethod project (vector1 vector2)
+  (dot vector1 (normalize vector2)))
+
+
+(defvar +float2-zero+ (make-instance 'float2))
+(defvar +float3-zero+ (make-instance 'float3))
+(defvar +float4-zero+ (make-instance 'float4))
+
+(defvar +float2-one+ (make-instance 'float2 :x 1 :y 1))
+(defvar +float3-one+ (make-instance 'float3 :x 1 :y 1 :z 1))
+(defvar +float4-one+ (make-instance 'float4 :x 1 :y 1 :z 1 :w 1))
+
+(defvar +float2-unit+ (normalize +float2-one+))
```

Comment:
Usually unit vectors are thought as that form current basis. So in 2D there are two unit vectors, three in 3D and so on.

Reply:
Done